### ISSUE TITLE
Grandpa panic

### ISSUE THREAD

**Issue by ascjones** [paritytech/substrate#5000]
Just got this panic running local `--dev` chain:

```
Thread 'main-tokio-' panicked at 'Receiver::next_message called after `None`', src/libcore/option.rs:1188
```

```
  10: <sc_finality_grandpa::communication::NetworkBridge<B,N> as core::future::future::Future>::poll
  11: <sc_finality_grandpa::VoterWork<B,E,Block,N,RA,SC,VR> as core::future::future::Future>::poll
  12: <futures_util::future::select::Select<A,B> as core::future::future::Future>::poll
```

Full backtrace:
```
Version: 2.0.0-992aea815-x86_64-linux-gnu

stack backtrace:
   0: sp_panic_handler::set::{{closure}}
   1: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:475
   2: rust_begin_unwind
             at src/libstd/panicking.rs:375
   3: core::panicking::panic_fmt
             at src/libcore/panicking.rs:84
   4: core::option::expect_failed
             at src/libcore/option.rs:1188
   5: futures_channel::mpsc::UnboundedReceiver<T>::next_message
   6: <futures_channel::mpsc::UnboundedReceiver<T> as futures_core::stream::Stream>::poll_next
   7: <futures_util::compat::compat03as01::Compat<St> as futures::stream::Stream>::poll
   8: futures::task_impl::std::set
   9: futures_util::future::future::FutureExt::poll_unpin
  10: <sc_finality_grandpa::communication::NetworkBridge<B,N> as core::future::future::Future>::poll
  11: <sc_finality_grandpa::VoterWork<B,E,Block,N,RA,SC,VR> as core::future::future::Future>::poll
  12: <futures_util::future::select::Select<A,B> as core::future::future::Future>::poll
  13: std::panicking::try::do_call
  14: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:78
  15: std::panicking::try
  16: <futures_util::future::select::Select<A,B> as core::future::future::Future>::poll
  17: <std::future::GenFuture<T> as core::future::future::Future>::poll
  18: tokio::task::core::Core<T>::poll
  19: std::panicking::try::do_call
  20: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:78
  21: tokio::task::harness::Harness<T,S>::poll
  22: tokio::runtime::thread_pool::worker::GenerationGuard::run_task
  23: tokio::runtime::thread_pool::worker::GenerationGuard::run
  24: std::thread::local::LocalKey<T>::with
  25: tokio::runtime::thread_pool::worker::Worker::run
  26: tokio::task::core::Core<T>::poll
  27: std::panicking::try::do_call
  28: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:78
  29: tokio::task::harness::Harness<T,S>::poll
  30: tokio::runtime::blocking::pool::Inner::run
  31: tokio::runtime::context::enter
  32: std::sys_common::backtrace::__rust_begin_short_backtrace
  33: std::panicking::try::do_call
  34: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:78
  35: core::ops::function::FnOnce::call_once{{vtable.shim}}
  36: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/liballoc/boxed.rs:1022
  37: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/5e1a799842ba6ed4a57e91f7ab9435947482f7d8/src/liballoc/boxed.rs:1022
      std::sys_common::thread::start_thread
             at src/libstd/sys_common/thread.rs:13
      std::sys::unix::thread::Thread::new::thread_start
             at src/libstd/sys/unix/thread.rs:80
  38: start_thread
  39: __clone
```

**Comment by mxinden** [comment]
Thanks for creating the issue @ascjones! Below is what I think is happening.

`NetworkBridge<B,N>::poll` does 3 things:

1. Poll the neighbor packet worker.
2. Poll the gossip validator report stream.
3. Poll the gossip engine.

Both polling the [neighbor packet worker](https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/communication/mod.rs#L430) as well as polling the [gossip validator report stream](https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/communication/mod.rs#L442) properly return on `Poll::Ready(None)`. This then bubbles up all the way to the `VoterWorker` and terminates the root future.

After (1) and (2) the `NetworkBridge` [polls the GossipEngine](https://github.com/paritytech/substrate/blob/master/client/finality-grandpa/src/communication/mod.rs#L453). Within the `GossipEngine` poll method [the network stream is polled](https://github.com/paritytech/substrate/blob/master/client/network-gossip/src/bridge.rs#L151), but `Poll::Ready(None)` return values are ignored. Thus on subsequent calls this would panic.

Now [polling the network stream](https://github.com/paritytech/substrate/blob/master/client/network-gossip/src/bridge.rs#L151) should only ever return `None` when either an `unbounded_send` fails, thus the [sender is dropped](https://github.com/paritytech/substrate/blob/master/client/network/src/service.rs#L968) or the entire `Network` is dropped. Both I would deem fairly unlikely.

Anyways, we should not be ignoring the `Poll::Ready(None)`. I will create a pull request.

**Comment by bkchr** [comment]
Was fixed by #5282
